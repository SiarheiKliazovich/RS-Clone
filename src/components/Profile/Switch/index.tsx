import "./switch.scss";
import { FunctionComponent } from "react";
import { useDispatch } from "react-redux";
import { ActionCreator, AnyAction } from "redux";

export type SwitchType = {
  action: ActionCreator<AnyAction>;
};

const Switch: FunctionComponent<SwitchType> = ({ action }: SwitchType) => {
  const dispatch = useDispatch();

  const updateState = (isChecked: boolean) => dispatch(action(isChecked));

  return (
    <label className="switch">
      <input
        className="switch__checkbox"
        type="checkbox"
        onClick={(e) => updateState((e.target as HTMLInputElement).checked)}
      />
      <span className="switch__slider"></span>
    </label>
  );
};

export default Switch;
