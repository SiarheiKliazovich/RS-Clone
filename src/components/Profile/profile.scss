@import "../../sass/mixins";
@import "../../sass/variables";

.profile {
  width: 350px;
  height: 100vh;
  display: flex;
  flex-direction: column;
  position: fixed;
  top: 0;
  right: 0;
  opacity: 0;
  background-color: $color-element-background-light;
  transition: $transition;
  transform: translateX(350px);
  box-shadow: $box-shadow;
  overflow: auto;
  z-index: 25;

  &--show {
    opacity: 1;
    transform: translateX(0);
  }
}

.profile__header {
  padding: 15px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  border-bottom: $border;
}

.profile__close {
  padding: 5px 10px;
  font-size: $font-size-xs;
}

.profile__wrapper {
  padding: 15px;
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  flex-grow: 2;
}

.profile__user {
  margin: 0 10px 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 10px;
}

.profile__image {
  @include icon-background('../../assets/images/jpg/user.jpg', 80px);
  border-radius: $border-radius-sm;
  cursor: pointer;
}

.profile__email {
  font-size: $font-size-llg;
  font-weight: $font-weight-bolder;
  line-height: $line-height-l;
  text-align: center;
}

.profile__role {
  font-size: $font-size-lg;
  text-transform: uppercase;
}

.profile__title,
.settings__title {
  font-size: $font-size-llg;
}

.settings__list {
  margin: 15px 0;
  display: flex;
  flex-direction: column;
  gap: 10px;

  &--row {
    flex-direction: unset;
    justify-content: space-evenly;
  }
}

.settings__item {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.settings__name {
  font-size: $font-size-lg;
}

.settings__colors {
  margin: 40px 0;
}

.settings__subtitle {
  font-size: $font-size-lg;
  text-align: center;
}

.settings__color {
  display: block;
  width: 25px;
  height: 25px;
  border-radius: 50%;
  background-color: $color-accent;
  transition: $transition;
  cursor: pointer;

  &:hover {
    background-color: $color-accent-hover;
  }

  &--forest {
    background-color: $color-accent-green;

    &:hover {
      background-color: $color-accent-green-hover;
    }
  }

  &--purple {
    background-color: $color-accent-purple;

    &:hover {
      background-color: $color-accent-purple-hover;
    }
  }

  &--coffee {
    background-color: $color-accent-coffee;

    &:hover {
      background-color: $color-accent-coffee-hover;
    }
  }
}

.settings__button {
  width: 100%;
}

.links__wrapper {
  display: flex;
  flex-direction: column;
  gap: 5px;

  .link {
    display: block;
    width: 100%;
    font-size: 18px;
    border-radius: 5px;
    text-align: center;
    color: $color-white1;
    padding: 3px 0;

    &:not(:last-child) {
      margin-bottom: 5px;
    }
  }
}

.button__link {
  text-align: center;
}
