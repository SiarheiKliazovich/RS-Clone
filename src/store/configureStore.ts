/* eslint-disable @typescript-eslint/ban-ts-comment */
import { applyMiddleware, createStore } from "redux";
import thunkMiddleware from "redux-thunk";

import loggerMiddleware from "./logger";
import rootReducer from "../reducers";

const configureStore = () => {
  const middlewares = [loggerMiddleware, thunkMiddleware];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(rootReducer, middlewareEnhancer);

  // @ts-ignore:next-line
  if (process.env.NODE_ENV !== "production" && module.hot) {
    // @ts-ignore:next-line
    module.hot.accept("./../reducers", () => store.replaceReducer(rootReducer));
  }

  return store;
};

export default configureStore;
