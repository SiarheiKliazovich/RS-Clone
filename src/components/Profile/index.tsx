import "./profile.scss";
import { FunctionComponent, useRef } from "react";
import { Link, useNavigate } from "react-router-dom";
import classNames from "classnames";
import Hover from "../Hover";
import { useDispatch, useSelector } from "react-redux";
import Switch from "./Switch";
import { ProfileType } from "../../types";
import { accentColors } from "../../data/constants";
import useOnClickOutside from "../../hook/useOnClickOutside";
import {
  updateFixedSidebar,
  updateNightMode,
  updateSidebarAccentMode,
  updateAccentColor,
  updateNavbarNightMode,
  showProfile,
  resetInterfaceToDefault,
} from "../../actions";
import { useTranslation } from "react-i18next";
import { IAuthReducer } from "../../interfaces/authReducer";
import { authorization } from "../../actions";
import { ILogin } from "../../interfaces/login";

const Profile: FunctionComponent<ProfileType> = ({
  accentColor,
  isProfileShow,
  isSidebarFixed,
  isSidebarAccentMode,
  isNavbarNightMode,
  isNightMode,
}: ProfileType) => {
  const dispatch = useDispatch();
  const ref = useRef<HTMLDivElement>(null);
  const navigate = useNavigate();

  const updateColor = (color: { static: string; hover: string }) =>
    dispatch(updateAccentColor(color));

  useOnClickOutside(ref, () => dispatch(showProfile()), isProfileShow);

  const { t } = useTranslation("dataLang");

  const styles = {
    backgroundColor: accentColor.static,
  };

  const authStore = useSelector((state: IAuthReducer) => state.auth);
  const { user, isLogin } = authStore;

  const handlerSign = () => {
    if (isLogin) {
      dispatch(authorization({} as ILogin, false));
    } else {
      navigate("/sign-in");
    }
  };

  return (
    <div
      className={classNames("profile", { "profile--show": isProfileShow })}
      ref={ref}
    >
      <div className="profile__header">
        <h2 className="profile__title">{t("profile.profile")}</h2>
        <Hover>
          <button
            className="button profile__close"
            onClick={() => dispatch(showProfile())}
          >
            X
          </button>
        </Hover>
      </div>
      <div className="profile__wrapper">
        <div className="profile__user">
          <span className="profile__image"></span>
          <p className="profile__email">{user.email}</p>
          <p className="profile__role">{user.role}</p>
          <Hover>
            <button className="button profile__sign-out" onClick={handlerSign}>
              {isLogin ? t("profile.signOut") : t("profile.signIn")}
            </button>
          </Hover>
        </div>
        {user.role === "USER" ? (
          <div className="links__wrapper">
            <Hover>
              <Link style={styles} className="link" to="/profile">
                {t("profile.myProfile")}
              </Link>
            </Hover>
            <Hover>
              <Link style={styles} className="link" to="/my-orders">
                {t("profile.myOrders")}
              </Link>
            </Hover>
          </div>
        ) : null}
        {user.role === "SELLER" ? (
          <div className="links__wrapper">
            <Hover>
              <Link style={styles} className="link" to="/profile">
                {t("profile.myProfile")}
              </Link>
            </Hover>
            <Hover>
              <Link style={styles} className="link" to="/my-ties">
                {t("profile.myTies")}
              </Link>
            </Hover>
          </div>
        ) : null}
        <div className="settings">
          <h2 className="settings__title">{t("profile.settings")}</h2>
          <ul className="settings__list">
            <li className="settings__item">
              <span className="settings__name">{t("profile.nightMode")}</span>
              <Switch
                action={updateNightMode}
                accentColor={accentColor}
                isChecked={isNightMode}
                isNightMode={isNightMode}
              />
            </li>
            <li className="settings__item">
              <span className="settings__name">
                {t("profile.sidebarFixed")}
              </span>
              <Switch
                action={updateFixedSidebar}
                accentColor={accentColor}
                isChecked={isSidebarFixed}
                isNightMode={isNightMode}
              />
            </li>
            <li className="settings__item">
              <span className="settings__name">
                {t("profile.sidebarAccentMode")}
              </span>
              <Switch
                action={updateSidebarAccentMode}
                accentColor={accentColor}
                isChecked={isSidebarAccentMode}
                isNightMode={isNightMode}
              />
            </li>
            <li className="settings__item">
              <span className="settings__name">
                {t("profile.navbarNightMode")}
              </span>
              <Switch
                action={updateNavbarNightMode}
                accentColor={accentColor}
                isChecked={isNavbarNightMode}
                isNightMode={isNightMode}
              />
            </li>
          </ul>
          <div className="settings__colors">
            <h4 className="settings__subtitle">
              {t("profile.changeAccentColor")}
            </h4>
            <ul className="settings__list settings__list--row">
              <li className="settings__item">
                <span
                  className="settings__color settings__color--default"
                  onClick={() => updateColor(accentColors.default)}
                ></span>
              </li>
              <li className="settings__item">
                <span
                  className="settings__color settings__color--forest"
                  onClick={() => updateColor(accentColors.forest)}
                ></span>
              </li>
              <li className="settings__item">
                <span
                  className="settings__color settings__color--purple"
                  onClick={() => updateColor(accentColors.purple)}
                ></span>
              </li>
              <li className="settings__item">
                <span
                  className="settings__color settings__color--coffee"
                  onClick={() => updateColor(accentColors.coffee)}
                ></span>
              </li>
            </ul>
          </div>
          <Hover>
            <button
              className="button settings__button"
              onClick={() => dispatch(resetInterfaceToDefault())}
            >
              {t("profile.reset")}
            </button>
          </Hover>
        </div>
      </div>
    </div>
  );
};

export default Profile;
