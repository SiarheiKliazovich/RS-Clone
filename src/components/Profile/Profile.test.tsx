import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import Profile from ".";

const mockStore = configureStore([]);

describe("Profile", () => {
  // eslint-disable-next-line
  let store: any;

  beforeEach(() => {
    store = mockStore({
      appInterface: {
        isNightMode: false,
        accentColor: {
          static: "#000",
          hover: "#fff",
        },
      },
      auth: {
        user: {
          role: "USER",
        },
        isLogin: true,
      },
    });
  });

  test("renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Profile
            accentColor={{ static: "#fff", hover: "#000" }}
            isProfileShow={true}
            isSidebarFixed={true}
            isSidebarAccentMode={true}
            isNavbarNightMode={false}
            isNightMode={false}
          />
        </Router>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
