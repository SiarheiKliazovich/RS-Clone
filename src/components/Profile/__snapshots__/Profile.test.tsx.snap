// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile renders correctly 1`] = `
<div>
  <div
    class="profile profile--show"
  >
    <div
      class="profile__header"
    >
      <h2
        class="profile__title"
      >
        profile.profile
      </h2>
      <button
        class="button profile__close"
        style="background-color: rgb(0, 0, 0);"
      >
        X
      </button>
    </div>
    <div
      class="profile__wrapper"
    >
      <div
        class="profile__user"
      >
        <span
          class="profile__image"
        />
        <p
          class="profile__email"
        />
        <p
          class="profile__role"
        >
          USER
        </p>
        <button
          class="button profile__sign-out"
          style="background-color: rgb(0, 0, 0);"
        >
          profile.signOut
        </button>
      </div>
      <div
        class="links__wrapper"
      >
        <a
          class="link"
          href="/profile"
          style="background-color: rgb(0, 0, 0);"
        >
          profile.myProfile
        </a>
        <a
          class="link"
          href="/my-orders"
          style="background-color: rgb(0, 0, 0);"
        >
          profile.myOrders
        </a>
        <a
          class="link"
          href="/favourite-tie"
          style="background-color: rgb(0, 0, 0);"
        >
          profile.favourite
        </a>
      </div>
      <div
        class="settings"
      >
        <h2
          class="settings__title"
        >
          profile.settings
        </h2>
        <ul
          class="settings__list"
        >
          <li
            class="settings__item"
          >
            <span
              class="settings__name"
            >
              profile.nightMode
            </span>
            <label
              class="switch"
            >
              <input
                class="switch__checkbox"
                type="checkbox"
              />
              <span
                class="switch__slider"
                style="background-color: rgb(255, 255, 255); border-color: #535f6b;"
              />
            </label>
          </li>
          <li
            class="settings__item"
          >
            <span
              class="settings__name"
            >
              profile.sidebarFixed
            </span>
            <label
              class="switch"
            >
              <input
                checked=""
                class="switch__checkbox"
                type="checkbox"
              />
              <span
                class="switch__slider"
                style="background-color: rgb(255, 255, 255); border-color: #fff;"
              />
            </label>
          </li>
          <li
            class="settings__item"
          >
            <span
              class="settings__name"
            >
              profile.sidebarAccentMode
            </span>
            <label
              class="switch"
            >
              <input
                checked=""
                class="switch__checkbox"
                type="checkbox"
              />
              <span
                class="switch__slider"
                style="background-color: rgb(255, 255, 255); border-color: #fff;"
              />
            </label>
          </li>
          <li
            class="settings__item"
          >
            <span
              class="settings__name"
            >
              profile.navbarNightMode
            </span>
            <label
              class="switch"
            >
              <input
                class="switch__checkbox"
                type="checkbox"
              />
              <span
                class="switch__slider"
                style="background-color: rgb(255, 255, 255); border-color: #535f6b;"
              />
            </label>
          </li>
        </ul>
        <div
          class="settings__colors"
        >
          <h4
            class="settings__subtitle"
          >
            profile.changeAccentColor
          </h4>
          <ul
            class="settings__list settings__list--row"
          >
            <li
              class="settings__item"
            >
              <span
                class="settings__color settings__color--default"
              />
            </li>
            <li
              class="settings__item"
            >
              <span
                class="settings__color settings__color--forest"
              />
            </li>
            <li
              class="settings__item"
            >
              <span
                class="settings__color settings__color--purple"
              />
            </li>
            <li
              class="settings__item"
            >
              <span
                class="settings__color settings__color--coffee"
              />
            </li>
          </ul>
        </div>
        <button
          class="button settings__button"
          style="background-color: rgb(0, 0, 0);"
        >
          profile.reset
        </button>
      </div>
    </div>
  </div>
</div>
`;
